File: features/core/planning-v3/validator/fingerprint.ts
import type { PlanningProblemV3 } from "@/features/core/planning-v3/types/problem"
import type { PlanningSolutionV3 } from "@/features/core/planning-v3/types/solution"

/**
 * Deterministic fingerprints for V3 problems and solutions.
 *
 * The point is comparability, not secrecy: two runs that produce the same
 * schedule must produce the same string, and any difference in the assignments
 * must change it.
 *
 * Two independently seeded 32-bit FNV-1a passes are concatenated into a 64-bit
 * hex digest. `Math.imul` keeps the multiply in 32-bit integer space, so the
 * result is exact, synchronous and free of any BigInt or crypto dependency.
 */

const FNV_PRIME_32 = 0x01000193

function hash32(value: string, seed: number): number {
  let digest = seed
  for (let index = 0; index < value.length; index++) {
    digest = Math.imul(digest ^ value.charCodeAt(index), FNV_PRIME_32)
  }
  return digest >>> 0
}

function hash64(value: string): string {
  const low = hash32(value, 0x811c9dc5)
  const high = hash32(value, 0x9e3779b9)
  return high.toString(16).padStart(8, "0") + low.toString(16).padStart(8, "0")
}

/** Canonical, order-stable text for a problem. */
export function fingerprintProblem(problem: PlanningProblemV3): string {
  const parts: string[] = [
    problem.version,
    String(problem.planningId),
    problem.sectorId,
    `${problem.period.start}..${problem.period.end}`,
    `step=${problem.timeStepMinutes}`,
    `rules=${JSON.stringify(problem.rules)}`,
    `objectives=${problem.objectives.join(",")}`,
  ]
  if (problem.sectors !== undefined) parts.push(`sectors=${JSON.stringify(problem.sectors)}`)
  for (const employee of [...problem.employees].sort(byId)) {
    // The daily bounds and the right to split belong here as much as the
    // contract does: they decide what a day may hold, so two rosters that
    // differ on them pose different questions. Only the display name is left
    // out — it changes nothing a solver can act on.
    parts.push(
      `E|${String(employee.id)}|${employee.contractMinutes}|${employee.workingDays.join(".")}|${employee.fixedRestDays.join(".")}|${employee.minimumDailyMinutes}|${employee.maximumDailyMinutes}|${employee.canOpen ? 1 : 0}${employee.canClose ? 1 : 0}${employee.canSplitShift ? 1 : 0}|${employee.maximumOpenings ?? "-"}|${employee.maximumClosings ?? "-"}${employee.allowedSectorIds === undefined ? "" : `|${employee.allowedSectorIds.join(".")}`}`
    )
  }
  for (const day of [...problem.days].sort((left, right) => left.date.localeCompare(right.date))) {
    parts.push(
      `D|${day.date}|${day.closed ? 1 : 0}|${day.opensAtMinutes ?? "-"}|${day.closesAtMinutes ?? "-"}|${day.budgetMinutes}${day.budgetMode === undefined ? "" : `|${day.budgetMode}`}`
    )
  }
  // Per-person, per-day availability — the field this function was blind to for
  // longest, and the most costly one to be blind to.
  //
  // A week where someone is AWAY and the same week where they are there shared
  // one identity. So did two weeks differing only in when a person may first
  // arrive, or how late they may stay. A perturbation campaign built on this
  // fingerprint silently lost three of its six axes: every scenario looked like
  // the baseline and was discarded as a duplicate.
  //
  // Nothing downstream can recover from that. A cached result, a comparison
  // between engines, a regression pinned to a fingerprint — each would be
  // answering about a different roster than the one asked about, and would look
  // right while doing it.
  for (const entry of [...problem.employeeDays].sort(byEmployeeDay)) {
    parts.push(
      `A|${String(entry.employeeId)}|${entry.date}|${entry.available ? 1 : 0}${entry.mandatory ? 1 : 0}${entry.fixedRest ? 1 : 0}|${entry.earliestStartMinutes}|${entry.latestEndMinutes}|${entry.maximumMinutes}`
    )
  }
  // Closing history — hashed ONLY when a balance is on, which is exactly when
  // it can change an answer. Hashing it unconditionally would give a new
  // identity to every week whose history moved, invalidating stored results for
  // problems that would still be solved identically.
  //
  // `rules` already carries the policy and is hashed whole, so switching a
  // balance on moves the digest on its own; this block adds the numbers the
  // engines will actually read.
  for (const entry of [...(problem.closingHistory ?? [])].sort(byEmployeeId)) {
    parts.push(
      `H${entry.sectorId === undefined ? "" : `|${entry.sectorId}`}|${String(entry.employeeId)}|${entry.closings}|${entry.opportunities}|${entry.saturdayClosings}|${entry.saturdayOpportunities}`
    )
  }
  for (const slot of [...problem.demandSlots].sort(bySlot)) {
    // The head-count fields are ALL of them, including the ones that may be
    // absent. A fingerprint blind to `hardMinimumEmployees` gave the same
    // identity to a week that must never be left unattended and a week where
    // the same shortfall is a degradation to accept — two different questions
    // wearing one name, which is the single thing this function exists to stop.
    // Absent is written as `-` rather than skipped, so "not declared" and
    // "declared zero" stay distinguishable.
    parts.push(
      `S|${slot.id}${slot.sectorId === undefined ? "" : `|${slot.sectorId}`}|${slot.date}|${slot.startMinutes}|${slot.endMinutes}|${slot.requiredEmployees}|${slot.hardMinimumEmployees ?? "-"}|${slot.maximumEmployees ?? "-"}`
    )
  }
  return `p3_${hash64(parts.join("\n"))}`
}

/** Canonical, order-stable text for a solution. */
export function fingerprintSolution(solution: PlanningSolutionV3): string {
  const rows = solution.assignments
    .map(
      (assignment) =>
        `${String(assignment.employeeId)}|${assignment.date}|${[...assignment.segments]
          .sort((left, right) => left.startMinutes - right.startMinutes)
          .map((segment) => `${segment.startMinutes}-${segment.endMinutes}`)
          .join(",")}${assignment.sectorAssignments === undefined ? "" : `|${[...assignment.sectorAssignments]
            .sort((left, right) => left.startMinutes - right.startMinutes)
            .map((block) => `${block.sectorId}:${block.startMinutes}-${block.endMinutes}`)
            .join(",")}`}`
    )
    .sort()
  return `s3_${hash64([solution.version, solution.problemFingerprint, ...rows].join("\n"))}`
}

function byId(left: { id: unknown }, right: { id: unknown }): number {
  return String(left.id).localeCompare(String(right.id))
}

function byEmployeeId(left: { employeeId: unknown }, right: { employeeId: unknown }): number {
  return String(left.employeeId).localeCompare(String(right.employeeId))
}

function byEmployeeDay(
  left: { employeeId: unknown; date: string },
  right: { employeeId: unknown; date: string }
): number {
  return (
    String(left.employeeId).localeCompare(String(right.employeeId)) ||
    left.date.localeCompare(right.date)
  )
}

function bySlot(
  left: { date: string; startMinutes: number; id: string },
  right: { date: string; startMinutes: number; id: string }
): number {
  return (
    left.date.localeCompare(right.date) ||
    left.startMinutes - right.startMinutes ||
    left.id.localeCompare(right.id)
  )
}

File: experiments/planning-v3-highs/shiftos_highs/fingerprint.py
from __future__ import annotations

import json
from typing import Any

FNV_PRIME_32 = 0x01000193


def _hash32(value: str, seed: int) -> int:
    digest = seed & 0xFFFFFFFF
    for char in value:
        digest = ((digest ^ ord(char)) * FNV_PRIME_32) & 0xFFFFFFFF
    return digest


def _hash64(value: str) -> str:
    low = _hash32(value, 0x811C9DC5)
    high = _hash32(value, 0x9E3779B9)
    return f"{high:08x}{low:08x}"


def _js_stringify(value: Any) -> str:
    # Matches JSON.stringify for the scalar/object shapes used in Planning V3.
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def fingerprint_problem(problem: dict[str, Any]) -> str:
    parts = [
        problem["version"],
        str(problem["planningId"]),
        problem["sectorId"],
        f'{problem["period"]["start"]}..{problem["period"]["end"]}',
        f'step={problem["timeStepMinutes"]}',
        f'rules={_js_stringify(problem["rules"])}',
        f'objectives={",".join(problem["objectives"])}',
    ]
    for employee in sorted(problem["employees"], key=lambda item: str(item["id"])):
        # The daily bounds and the right to split decide what a day may hold, so
        # two rosters differing on them pose different questions. Only the
        # display name is left out — nothing acts on it.
        parts.append(
            f'E|{employee["id"]}|{employee["contractMinutes"]}|'
            f'{".".join(employee["workingDays"])}|{".".join(employee["fixedRestDays"])}|'
            f'{employee["minimumDailyMinutes"]}|{employee["maximumDailyMinutes"]}|'
            f'{1 if employee["canOpen"] else 0}{1 if employee["canClose"] else 0}'
            f'{1 if employee["canSplitShift"] else 0}|'
            f'{employee["maximumOpenings"] if employee["maximumOpenings"] is not None else "-"}|'
            f'{employee["maximumClosings"] if employee["maximumClosings"] is not None else "-"}'
        )
    for day in sorted(problem["days"], key=lambda item: item["date"]):
        parts.append(
            "D|{}|{}|{}|{}|{}".format(
                day["date"],
                1 if day["closed"] else 0,
                day["opensAtMinutes"] if day["opensAtMinutes"] is not None else "-",
                day["closesAtMinutes"] if day["closesAtMinutes"] is not None else "-",
                day["budgetMinutes"],
            )
        )
    # Per-person, per-day availability. The field both sides were blind to for
    # longest, and the most costly one to be blind to: a week where someone is
    # AWAY and the same week where they are there shared one identity, as did
    # two weeks differing only in when a person may arrive or leave. A
    # perturbation campaign built on the old fingerprint silently lost three of
    # its six axes — every scenario looked like the baseline.
    for entry in sorted(
        problem["employeeDays"], key=lambda item: (str(item["employeeId"]), item["date"])
    ):
        parts.append(
            f'A|{entry["employeeId"]}|{entry["date"]}|'
            f'{1 if entry["available"] else 0}{1 if entry["mandatory"] else 0}'
            f'{1 if entry["fixedRest"] else 0}|'
            f'{entry["earliestStartMinutes"]}|{entry["latestEndMinutes"]}|'
            f'{entry["maximumMinutes"]}'
        )
    for slot in sorted(
        problem["demandSlots"],
        key=lambda item: (item["date"], item["startMinutes"], item["id"]),
    ):
        # Mirrors the TypeScript fingerprint exactly, INCLUDING the head-count
        # fields that may be absent. Both sides used to omit
        # `hardMinimumEmployees`, which gave one identity to a week that must
        # never be left unattended and a week where the same shortfall is only a
        # degradation. Absent is written as `-` rather than skipped, so "not
        # declared" and "declared zero" stay distinguishable.
        hard = slot.get("hardMinimumEmployees")
        maximum = slot.get("maximumEmployees")
        parts.append(
            f'S|{slot["id"]}|{slot["date"]}|{slot["startMinutes"]}|{slot["endMinutes"]}|'
            f'{slot["requiredEmployees"]}|{hard if hard is not None else "-"}|'
            f'{maximum if maximum is not None else "-"}'
        )
    return f'p3_{_hash64(chr(10).join(parts))}'


def fingerprint_solution(solution: dict[str, Any]) -> str:
    rows: list[str] = []
    for assignment in solution["assignments"]:
        segments = sorted(assignment["segments"], key=lambda item: item["startMinutes"])
        encoded = ",".join(
            f'{segment["startMinutes"]}-{segment["endMinutes"]}' for segment in segments
        )
        rows.append(f'{assignment["employeeId"]}|{assignment["date"]}|{encoded}')
    rows.sort()
    value = "\n".join([solution["version"], solution["problemFingerprint"], *rows])
    return f's3_{_hash64(value)}'
